### DISCUSSION TITLE
Don't automatically cast all env-vars to numbers

### DISCUSSION THREAD

**Discussion by cupcakearmy** [directus/directus#24893] [Feature Requests]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

I'm trying to get openid to work but no luck.

```yaml
version: '3.8'

services:
  db:
    image: postgres:14-alpine
    restart: unless-stopped
    env_file: .env
    volumes:
      - ./data/db:/var/lib/postgresql/data

  cache:
    image: redis:6-alpine
    restart: unless-stopped

  directus:
    image: directus/directus:9
    restart: unless-stopped
    env_file: .env
    volumes:
      - ./data/uploads:/directus/uploads
    depends_on:
      - cache
      - db
    ports:
      - 80:8055
```

```
POSTGRES_USER=test
POSTGRES_DB=test
POSTGRES_PASSWORD=***

KEY=***
SECRET=***
ACCESS_TOKEN_TTL=1d
REFRESH_TOKEN_TTL=30d

PUBLIC_URL=http://localhost

AUTH_PROVIDERS="slack"

AUTH_SLACK_DRIVER="openid"
AUTH_SLACK_CLIENT_ID="***"
AUTH_SLACK_CLIENT_SECRET="***"
AUTH_SLACK_ISSUER_URL="https://slack.com/.well-known/openid-configuration"
AUTH_SLACK_ALLOW_PUBLIC_REGISTRATION=true

DB_CLIENT=pg
DB_HOST=db
DB_PORT=5432
DB_DATABASE=test
DB_USER=test
DB_PASSWORD=***

CACHE_ENABLED=true
CACHE_STORE=redis
CACHE_REDIS=redis://cache:6379

ADMIN_EMAIL=admin@example.org
ADMIN_PASSWORD=changeme
```

I've tested the OpenID flow with https://openidconnect.net/ and it works.

What is interesting, [looking at the code](https://github.com/directus/directus/blob/28ac068720f63da5ee21911a738a16d24fd7ef21/api/src/auth/drivers/openid.ts#L25-L41)

It should be an error before if `clientId` would not be defined [here](https://github.com/directus/directus/blob/28ac068720f63da5ee21911a738a16d24fd7ef21/api/src/auth/drivers/openid.ts#L28)

However the error comes from the `openid-client` library.
Am I missing something?
Thanks & I hope it's not a duplicate, could not find anything

### To Reproduce

Get the latest release `v9.0.0` and try slack oauth or openid

### Errors Shown

Stacktrace:

```
directus-directus-1  | 12:37:25 ✨ Initializing bootstrap...
directus-directus-1  | 12:37:25 ✨ Database already initialized, skipping install
directus-directus-1  | 12:37:25 ✨ Running migrations...
directus-directus-1  | 12:37:25 ✨ Done
directus-directus-1  | npm notice 
directus-directus-1  | npm notice New patch version of npm available! 8.1.0 -> 8.1.3
directus-directus-1  | npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.1.3>
directus-directus-1  | npm notice Run `npm install -g npm@8.1.3` to update!
directus-directus-1  | npm notice 
directus-directus-1  | 12:37:28 ⚠️  PostGIS isn't installed. Geometry type support will be limited.
directus-directus-1  | 12:37:29 ✨ Server started at http://localhost:8055
directus-directus-1  | /directus/node_modules/openid-client/lib/client.js:173
directus-directus-1  |       throw new TypeError('client_id is required');
directus-directus-1  |             ^
directus-directus-1  | 
directus-directus-1  | TypeError: client_id is required
directus-directus-1  |     at new BaseClient (/directus/node_modules/openid-client/lib/client.js:173:13)
directus-directus-1  |     at new Client (/directus/node_modules/openid-client/lib/client.js:1742:7)
directus-directus-1  |     at /directus/node_modules/directus/dist/auth/drivers/openid.js:34:25
directus-directus-1  |     at processTicksAndRejections (node:internal/process/task_queues:96:5)
directus-directus-1 exited with code 1
```

### What version of Directus are you using?

v9.0.0

### What version of Node.js are you using?

The one in the official docker image

### What database are you using?

postgres:14-alpine

### What browser are you using?

FF

### What operating system are you using?

macOS

### How are you deploying Directus?

Docker

**Comment by rijkvanzanten** [comment]
Odd! I just copy pasted your example (including the `***`), and I'm not seeing that error on startup 🤔 

Could there be a discrepancy in how those env vars are loaded into Docker? One thing I noticed is that the env vars don't use quotes, except for the AUTH ones, maybe that doesn't fly with Docker's way of loading in the .env file?

**Comment by rijkvanzanten** [comment]
(@danilopolani this is exactly the sort of config gotcha that prevents us from doing cool things like in your PR 😄 )

**Comment by cupcakearmy** [comment]
Took me a good 30 min but I got it:
Turn out Slack Client IDs are a long string of digits and have a `.` inside. (e.g. xxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx)
Having a dot inside of the client id causes the break. Why?

Well what happens is that [here](https://github.com/directus/directus/blob/b7ee318e4329f3623c41b222b6ba85433b8714ca/api/src/env.ts#L264-L272) directus thinks it's a number because of the `isNan` function and casts it to a number. Now since the leading number is quite big, decimal places are lost, altering the value. I guess from then the openid-client library only accepts a string and not a number

So I don't know how much numbers are required in the config file, so can't tell if this is a required feature, however makes oauth/openid not usable with slack.

**Comment by batuhanbilginn** [comment]
Yeah, it's the same for Facebook Client ID.

**Comment by rijkvanzanten** [comment]
You should be able to force cast it to a string by using `AUTH_SLACK_CLIENT_ID="string:xxx.xxx.xxx"`

We'll leave this open, as I agree we should probably not cast things to numbers by default, and instead rely on a hardcoded list of env vars that we expect to be numbers 👍🏻

**Comment by cupcakearmy** [comment]
Can confirm that with `"string:..." it works. Maybe a small "watch out" section in the docs would be helpfull :)
Thanks for the quick replies as always ❤️

**Comment by rijkvanzanten** [comment]
Linear: [ENG-287](https://linear.app/directus/issue/ENG-287/dont-automatically-cast-all-env-vars-to-numbers)

**Comment by ched-dev** [comment]
I just ran into this issue with facebook OAuth login. They use all numbers for their CLIENT_ID which was giving me an error:
> TypeError: client_id is required

I'm using Directus 10.3.0

**Comment by robsoncombr** [comment]
Hello, i'm experiencing similar issue and looking for pre-existent reports about it I found this one coming from #23645 

I also found other references around similar problem.
#11956 > [https://github.com/panva/openid-client/pull/477](https://github.com/panva/openid-client/pull/477)

So I first confirmed this is really affecting numbers only, I tried to set only the id _=11111_, with _='11111'_ and " _="11111"_, any try fails the same way, so I tried with _=a11111_ and it bypassed the error.

I have decided to apply a temporary change on my code while this is under discussion and implementation.

I changed [https://github.com/directus/directus/blob/900bf92d8733776d185fe4811eaab6f89d43d3c3/api/src/auth/drivers/oauth2.ts#L81](https://github.com/directus/directus/blob/900bf92d8733776d185fe4811eaab6f89d43d3c3/api/src/auth/drivers/oauth2.ts#L81)

with:
`client_id: clientId.toString(),
`

Since this is currently affecting only OAuth2 I will not touch the other drivers, I will also not post any PR since this is under discussion and involves more, but if anyone want this working for all OAuth2 Providers, this is a trick until we get a permanent solution.

0,02 -- cheers!

**Comment by robsoncombr** [comment]
Just added the patch to my codebase as a temporary patch, if you need this check it out.

[https://github.com/robsoncombr/directus/commit/24a57f3f912ac279a0cdaee27c5d0b5f998c1a47](https://github.com/robsoncombr/directus/commit/24a57f3f912ac279a0cdaee27c5d0b5f998c1a47)

**Comment by robsoncombr** [comment]
hi there, I have just found this into the documentation:

[https://docs.directus.io/self-hosted/config-options.html#type-casting-and-nesting](https://docs.directus.io/self-hosted/config-options.html#type-casting-and-nesting)

Then I found

[https://docs.directus.io/self-hosted/config-options.html#environment-syntax-prefix](https://docs.directus.io/self-hosted/config-options.html#environment-syntax-prefix)

So, basically we can by-pass the casting and provide the type we need, it works to me and I was not aware, I will test later so I can revert my temp patch.

I let you know.

**Comment by rijkvanzanten** [comment]
@robsoncombr That doc is the recommended workaround! This "bug" was implemented as a feature, to make the env vars a bit more friendly to setup. It does result in some unexpected behavior however — like an oauth2 id accidentally being seen as a number instead of a string. The proper 'fix' here is to remove the auto-casting feature altogether, but that is quite the breaking change as it would require folks to update their .env on upgrade. It's on the backlog, but not something I'm looking to ship until we do the next major version release.

I'm trying to keep patches and minors as backwards compatible as possible, and collect breaking features and breaking changes that require a manual migration into bigger major releases that we only ship every couple of years. The last one of those was v8-v9. We're about due for the next one 🙂
